src/main/cpp/src/GpuTimeZoneDBJni.cpp

```diff
@@ -50,4 +50,50 @@ Java_com_nvidia_spark_rapids_jni_GpuTimeZoneDB_convertUTCTimestampColumnToTimeZo
   }
   CATCH_STD(env, 0);
 }
+
+JNIEXPORT jlong JNICALL
+Java_com_nvidia_spark_rapids_jni_GpuTimeZoneDB_timeAddCS(JNIEnv* env,
+                                                         jclass,
+                                                         jlong input_handle,
+                                                         jlong duration_handle,
+                                                         jlong transitions_handle,
+                                                         jint tz_index)
+{
+  JNI_NULL_CHECK(env, input_handle, "column is null", 0);
+  JNI_NULL_CHECK(env, transitions_handle, "column is null", 0);
+  try {
+    cudf::jni::auto_set_device(env);
+    auto const input = reinterpret_cast<cudf::column_view const*>(input_handle);
+    auto const duration =
```

Comment:
This is problematic because c++ is going to say that `duration_handle` is a pointer to `cudf::duration_scalar<cudf::duration_us>`, even if it is not. The only thing that is guaranteed is.

```
    auto duration = reinterpret_cast<cudf::scalar *>(duration_handle);
```

Beyond that you will need to use dynamic_cast where C++ will verify that it is the type you expect. Note that dynamic_cast returns nullptr if it is not the right type.

Reply:
Thanks. I moved the `dynamic_cast` and type checking to timezones.cu L392-L386 because scalar is forward declaration here. I'm not quite sure I did things right there, please help check again.